File: yw/dom/DocumentType.lua
--[[
    Copyright (c) 2024, Oh Inseo (YJK) <dhdlstjtr@gmail.com>
    SPDX-License-Identifier: BSD-3-Clause
    This software may contain third-party material. For more info, see README.
]]
local Node   = require "yw.dom.Node"
local object = require "yw.common.object"


---https://dom.spec.whatwg.org/#concept-doctype
---@class DOM_DocumentType : DOM_Node
---@field name string      https://dom.spec.whatwg.org/#concept-doctype-name
---@field publicID string  https://dom.spec.whatwg.org/#concept-doctype-publicid
---@field systemID string  https://dom.spec.whatwg.org/#concept-doctype-systemid
local DocumentType = object.create(Node)

---@param document DOM_Document
---@param name string
---@param publicID string?
---@param systemID string?
---@return DOM_DocumentType
function DocumentType:new(document, name, publicID, systemID)
    local o = Node.new(self, document) --[[@as DOM_DocumentType]]
    o.isDocumentType = true
    -- When a doctype is created, its name is always given.
    o.name = name
    -- Unless explicitly given when a doctype is created, its public ID and system ID are the empty string.
    o.publicID = publicID or ""
    o.systemID = systemID or ""
    return o
end

return DocumentType

File: dom/document_type.go
// This file is part of YW project. Copyright 2025 Oh Inseo (YJK)
// SPDX-License-Identifier: BSD-3-Clause
// See LICENSE for details, and LICENSE.thirdparty.md for third-party license information.

package dom

import (
	"fmt"
	"strings"
)

// DocumentType represents a [DOM doctype](a.k.a DocumentFragment).
//
// [DOM doctype]: https://dom.spec.whatwg.org/#concept-doctype
type DocumentType interface {
	Node

	// Name returns [name] of the doctype.
	//
	// [name]: https://dom.spec.whatwg.org/#concept-doctype-name
	Name() string

	// PublicId returns [public ID] of the doctype.
	//
	// [public ID]: https://dom.spec.whatwg.org/#concept-doctype-publicid
	PublicId() string

	// SystemId returns [system ID] of the doctype.
	//
	// [system ID]: https://dom.spec.whatwg.org/#concept-doctype-systemid
	SystemId() string
}
type documentTypeImpl struct {
	Node
	name     string
	publicId string
	systemId string
}

// NewDocumentType constructs a new [DocumentType] node.
func NewDocumentType(doc Document, name, publicId, systemId string) DocumentType {
	return documentTypeImpl{
		NewNode(doc),
		name, publicId, systemId,
	}
}
func (dt documentTypeImpl) String() string {
	sb := strings.Builder{}
	sb.WriteString("<!DOCTYPE")
	if dt.name != "" {
		sb.WriteString(fmt.Sprintf(" %s", dt.name))
	}
	if dt.publicId != "" {
		sb.WriteString(fmt.Sprintf(" PUBLIC %s", dt.publicId))
	}
	if dt.systemId != "" {
		sb.WriteString(fmt.Sprintf(" SYSTEM %s", dt.systemId))
	}
	sb.WriteString(">")
	return sb.String()
}
func (dt documentTypeImpl) Name() string     { return dt.name }
func (dt documentTypeImpl) PublicId() string { return dt.publicId }
func (dt documentTypeImpl) SystemId() string { return dt.systemId }
